src/librustc_mir/borrow_check/mutability_errors.rs

```diff
@@ -393,6 +395,63 @@ impl<'a, 'gcx, 'tcx> MirBorrowckCtxt<'a, 'gcx, 'tcx> {
                 );
             }
 
+            Place::Projection(box Projection {
+                base: Place::Local(local),
+                elem: ProjectionElem::Deref,
+            })  if error_access == AccessKind::MutableBorrow => {
+                err.span_label(span, format!("cannot {ACT}", ACT = act));
+
+                let mpi = self.move_data.rev_lookup.find_local(*local);
+                for i in self.move_data.init_path_map[mpi].iter() {
+                    if let InitLocation::Statement(location) = self.move_data.inits[*i].location {
+                        if let Some(
+                            Terminator {
+                                kind: TerminatorKind::Call {
+                                    func: Operand::Constant(box Constant {
+                                        literal: Const {
+                                            ty: &TyS {
+                                                sty: TyKind::FnDef(id, substs),
+                                                ..
+                                            },
+                                            ..
+                                        },
+                                        ..
+                                    }),
+                                    ..
+                                },
+                                ..
+                            }
+                        ) = &self.mir.basic_blocks()[location.block].terminator {
+                            if self.tcx.parent(id) == self.tcx.lang_items().index_trait() {
+
+                                let mut found = false;
+                                self.tcx.for_each_relevant_impl(
```

Comment:
Hmm, this check is pretty crude. I guess it'll work, but (e.g.) if `IndexMut` *is* implemented, but for a different index type, it will get the wrong result (though that is likely to get a different error, earlier in the compiler).

Another option might be to use the [`predicate_may_hold`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/infer/struct.InferCtxt.html#method.predicate_may_hold) method (or perhaps `predicate_must_hold`) to test if the trait holds.

It's a bit more work though — we'd have to construct the proper predicate and everything. Seems like something where some helper functions would be useful, I feel like this comes up from time to time in the code.

Reply:
We've discovered that due to a bug in the canonicalization code, it's not possible to switch to `predicate_may_hold` at the moment - the code that demonstrated this [is available on this branch](https://github.com/davidtwco/rust/tree/issue-53228-temp-predicate).